proposed/2023/allow-version-property-in-projectreference.md

```diff
@@ -0,0 +1,107 @@
+# Allow users to define version ranges for ProjectReferences
+
+- [Martin Ruiz](https://github.com/martinrrm)
+- Start Date (2023-01-01)
+- [5556](https://github.com/NuGet/Home/issues/5556)
+
+# Summary
+
+Add a `Version` to `ProjectReference` tag in CSPROJ, to allow customers to specify the referenced project version in the `.nupkg` and `nuspec` files when doing a pack command.
+
+# Motivation
+
+When using `ProjectReference` there is no option to define a Version to the reference like in `PackageReference` and the version will always be defined as `>= ProjectVersion`, this results in customers manually modifying the nuspec file if they want to declare a different version than the project version.
+
+# Explanation
+
+Currently when adding a `ProjectReference` to a project, there is no property to specify which version(s) of it to be used when doing a pack. 
+When doing a pack to a package with a `ProjectRefernce` it will always be added as a range, where the minumum version will be the ProjectReference version and with an open maximum version.
+
+```
+<ItemGroup>
+    <ProjectReference Include="../MyReferencedPackage/MyReferencedPackage.csproj" />
+</ItemGroup>
+```
+
+Add a `Version` property to `ProjectReference` and store that value in the assets file when doing a restore so we can retrieve that information when doing a `pack command.
+
+If there is no `Version` information then the behavior should be the current one.
+
+## Example
+
+### .CSPROJ file
+```
+<ItemGroup>
+    <PackageReference Include="Newtonsoft.Json" Version="[9.0.0, 13.0.2)" />
+    <ProjectReference Include="../MyReferencedPackage/MyReferencedPackage.csproj" Version="[1.0.0, 2.0.0)" />
```

Comment:
I think the minimum version should not be allowed to be specified on the ProjectReference, and NuGet should always automatically fill in the min version from the project's version.

This might not be popular with customers who want this feature, but this is for the protection of package consumers, especially since NuGet only selected minimum version in its dependency resolution graph.

Reply:
With this I think we can take the customer suggestion and do something like this:

```
<ProjectReference Include="..\MyReferencedPackage\MyReferencedPackage.csproj" >
    <MaximumVersion Inclusive="false">2</MaximumVersion>
</ProjectReference>
```

With this in mind, you think saving this as a version range in the assets file (like the initial proposal) or change it to be different properties?


Option 1 (I like this one more) :
```
"C:\\Users\\mruizmares\\source\\repos\\ConsoleApp1\\MyReferencedPackage\\MyReferencedPackage.csproj": {
  "projectPath": "C:\\Users\\mruizmares\\source\\repos\\ConsoleApp1\\MyReferencedPackage\\MyReferencedPackage.csproj",
  "version": "[1.0.0, 2.0.0)"
}
```


Option 2:
```
"C:\\Users\\mruizmares\\source\\repos\\ConsoleApp1\\MyReferencedPackage\\MyReferencedPackage.csproj": {
  "projectPath": "C:\\Users\\mruizmares\\source\\repos\\ConsoleApp1\\MyReferencedPackage\\MyReferencedPackage.csproj",
  "maxVersion": "2.0.0",
  "inclusive": true,
}
```


I currently have an mvp for this in branch `dev-martinrrm-project-reference` I have also tested with `dotnet build --no-restore` and it works, I'm going to investigate why is not breaking.

Reply:
Apparently I forgot to reply to this 😕 

> With this I think we can take the customer suggestion and do something like this:
> `<MaximumVersion Inclusive="false">2</MaximumVersion>`

That's not valid MSBuild syntax. While project files are XML, MSBuild has a strict syntax, and diverging from it will require changes to MSBuild if we feel stongly enough that it's needed to support a customer scenario (in which case you'd also need to consider what potential bugs it could introduce to non-NuGet scenarios if the syntax was extended). Anyway, MSBuild items are basically `Dictionary<string, string>`, where using XML attributes and child elements are equivalent. So any syntax proposal needs to be key-value pairs.

NuGet's `VersionRange` already supports values without a minimum value, for example `(, 2.0.0-0)`, so we could just have `<ProjectReference Include="..\path\to\project.csproj" Version="(,2.0.0)" />`, and then at pack time, NuGet figures out the actual version of `project.csproj`, let's say 1.2.3, and then makes it a `[1.2.3, 2.0.0)` dependency.

However, I think another scenario that some customers may want, is to limit the package dependency to the exact version of the project when it's being packed. VersionRange defines this as `[1.2.3]` (which is equivalent to `[1.2.3, 1.2.3]`). However, `[]` is not valid, neither is `[,]`. I don't currently have any ideas for syntax we can use, unless we special case `[]` in the relevant code before it gets passed to `VersionRange.Parse`.

Reply:
@zivkan I like the idea of doing `<ProjectReference Include="..\path\to\project.csproj" Version="(,2.0.0)" />` and change the version range to always have the lower version to the project version, just like your example. But at the same time I'm concern that this will only work if they are using ranges, otherwise it will look weird, for example:

`<ProjectReference Include="..\path\to\project.csproj" Version="2.0.0" />` make it a `[1.2.3, 2.0.0)` or `[1.2.3, 2.0.0]`.


What if we do something like `<ProjectReference Include="..\path\to\project.csproj" MaxVersion="2.0.0" Inclusive="true" />`? Is that a valid MSBuild syntax? Whit this we can also make the scenario for `[1.2.3,1.2.3]` work.

Reply:
magic word maybe?

`[X,1.2.3]` where X is the autofilled version?
Likewise, `[X]`.

Reply:
Today, the `Version` for project references get recalculated when pack is run, so maybe that's another argument for some sort of a magic word/letter. 

I don't love it though. It feels like the most completely way without misuse risk, but it's kind of ugly.